protocols/gossipsub/src/layer.rs

```diff
@@ -0,0 +1,1183 @@
+        if let Some(peer_list) = self.topic_peers.get(topic_hash) {
+            for peer in peer_list.floodsub.iter().chain(peer_list.gossipsub.iter()) {
+                debug!("Sending UNSUBSCRIBE to peer: {:?}", peer);
+                self.events.push_back(NetworkBehaviourAction::SendEvent {
+                    peer_id: peer.clone(),
+                    event: GossipsubRpc {
+                        messages: Vec::new(),
+                        subscriptions: vec![GossipsubSubscription {
+                            topic_hash: topic_hash.clone(),
+                            action: GossipsubSubscriptionAction::Unsubscribe,
+                        }],
+                        control_msgs: Vec::new(),
+                    },
+                });
+            }
+        }
+
+        // call LEAVE(topic)
+        // this will remove the topic from the mesh
+        self.leave(&topic_hash);
+
+        info!("Unsubscribed from topic: {:?}", topic_hash);
+        true
+    }
+
+    /// Publishes a message to the network.
+    pub fn publish(&mut self, topic: impl Into<TopicHash>, data: impl Into<Vec<u8>>) {
+        self.publish_many(iter::once(topic), data)
+    }
+
+    /// Publishes a message with multiple topics to the network.
+    pub fn publish_many(
+        &mut self,
+        topic: impl IntoIterator<Item = impl Into<TopicHash>>,
+        data: impl Into<Vec<u8>>,
+    ) {
+        let message = GossipsubMessage {
+            source: self.local_peer_id.clone(),
+            data: data.into(),
+            // If the sequence numbers are predictable, then an attacker could flood the network
+            // with packets with the predetermined sequence numbers and absorb our legitimate
+            // messages. We therefore use a random number.
+            // TODO: Check if the random sequence numbers causes issues with other clients.
+            // To be interoperable with the go-implementation this is treated as a 64-bit
+            // big-endian uint.
+            sequence_number: rand::random::<[u8; 8]>().to_vec(),
+            topics: topic.into_iter().map(|t| t.into().clone()).collect(),
+        };
+
+        debug!("Publishing message: {:?}", message.id());
+
+        // forward the message to mesh and floodsub peers
+        let local_peer_id = self.local_peer_id.clone();
+        self.forward_msg(message.clone(), local_peer_id);
+
+        let mut recipient_peers = HashMap::new();
+        for topic_hash in &message.topics {
+            // if not subscribed to the topic, use fanout peers
+            if self.mesh.get(&topic_hash).is_none() {
+                debug!("Topic: {:?} not in the mesh", topic_hash);
+                // build a list of peers to forward the message to
+                // if we have fanout peers add them to the map
+                if self.fanout.contains_key(&topic_hash) {
+                    for peer in self.fanout.get(&topic_hash).expect("Topic must exist") {
```

Comment:
Yes I will make these changes.

Reply:
It seems I did it this way because if I use 
```rust
if let some(peers) = self.fanout.get(&topic_hash) {
```
This borrows `self.fanout` and I then cannot insert into self.fanout in the case that there are no peers. i.e. the line
```rust
self.fanout.insert(topic_hash.clone(), new_peers.clone());
```